### ISSUE TITLE
openshift_master_identity_providers for OpenID config

### ISSUE THREAD

**Issue by anweiss** [openshift/openshift-ansible#1587]
Was wondering if there was an example showing how the `openshift_master_identity_providers` variable could be set for OpenID per settings [here](https://docs.openshift.org/latest/install_config/configuring_authentication.html#OpenID).

**Comment by detiber** [comment]
@anweiss assuming using an ansible inventory file (like inventory/byo/hosts.ose.example):
```
openshift_master_identity_providers=[{'name': 'htpasswd_auth', 'login': 'true', 'challenge': 'false', 'kind': 'OpenIDIdentityProvider', 'client_id': 'my_client_id', 'client_secret': 'my_client_secret', 'claims': { 'id': ['sub'], 'preferredUsername': ['preferred_username'], 'name': ['name'], 'email': ['email']}, 'urls': {'authorize': 'https://myidp.example.com/oauth2/authorize', 'token':  https://myidp.example.com/oauth2/token'}}]
```

If using a group_vars or host_vars file, you can get away with defining it in yaml:
```
openshift_master_identity_providers:
- name: my_openid_connect 
  challenge: false 
  login: true 
  mappingMethod: claim 
  kind: OpenIDIdentityProvider
  clientID: ... 
  clientSecret: ... 
  claims:
    id:
    - sub 
    preferredUsername:
    - preferred_username
    name:
    - name
    email:
    - email
  urls:
    authorize: https://myidp.example.com/oauth2/authorize 
    token: https://myidp.example.com/oauth2/token
```

As an aside, if you want to allow password based auth through the cli tooling, there will need to be an additional identity provider configured to handle for challenge auth, since the OpenID provider does not support challenge auth

**Comment by anweiss** [comment]
So, tried setting the variable using your example but getting the following error:

```
fatal: [osmaster01] => Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/ansible/runner/__init__.py", line 586, in _executor
    exec_rc = self._executor_internal(host, new_stdin)
  File "/usr/lib/python2.7/site-packages/ansible/runner/__init__.py", line 789, in _executor_internal
    return self._executor_internal_inner(host, self.module_name, self.module_args, inject, port, complex_args=complex_args)
  File "/usr/lib/python2.7/site-packages/ansible/runner/__init__.py", line 1013, in _executor_internal_inner
    complex_args = template.template(self.basedir, complex_args, inject, fail_on_undefined=self.error_on_undefined_vars)
  File "/usr/lib/python2.7/site-packages/ansible/utils/template.py", line 140, in template
    d[k] = template(basedir, v, templatevars, lookup_fatal, depth, expand_lists, convert_bare, fail_on_undefined, filter_fatal)
  File "/usr/lib/python2.7/site-packages/ansible/utils/template.py", line 124, in template
    varname = template_from_string(basedir, varname, templatevars, fail_on_undefined)
  File "/usr/lib/python2.7/site-packages/ansible/utils/template.py", line 382, in template_from_string
    res = jinja2.utils.concat(rf)
  File "<template>", line 9, in root
  File "/home/user/openshift-ansible/filter_plugins/openshift_master.py", line 476, in translate_idps
    IdentityProviderBase.validate_idp_list(idp_list)
  File "/home/user/openshift-ansible/filter_plugins/openshift_master.py", line 87, in validate_idp_list
    idp.validate()
  File "/home/user/openshift-ansible/filter_plugins/openshift_master.py", line 375, in validate
    if 'extraScopes' not in self.provider['extraScopes'] and not isinstance(self.provider['extraScopes'], list):
KeyError: 'extraScopes'
```

My config is as follows:

```
openshift_master_identity_providers=[{'name': 'azureadopenid', 'login': 'true', 'challenge': 'false', 'kind': 'OpenIDIdentityProvider', 'client_id': 'my_client_id', 'client_secret': 'my_client_secret', 'claims': { 'id': ['sub'], 'preferredUsername': ['unique_name'], 'name': ['unique_name'], 'email': ['unique_name']}, 'urls': {'authorize': 'authorize_url', 'token': 'token_url'}}]
```

**Comment by detiber** [comment]
>  File "/home/user/openshift-ansible/filter_plugins/openshift_master.py", line 375, in validate
>    if 'extraScopes' not in self.provider['extraScopes'] and not isinstance(self.provider['extraScopes'], list):
>KeyError: 'extraScopes'

oh, wow...  that should be `if 'extraScopes' in self.provider and not isinstance(self.provider['extraScopes'], list):`

You could work around it by adding `'extra_scopes': []`

**Comment by anweiss** [comment]
Happens when `extraAuthorizeParameters` is missing as well:

```
fatal: [osmaster01] => Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/ansible/runner/__init__.py", line 586, in _executor
    exec_rc = self._executor_internal(host, new_stdin)
  File "/usr/lib/python2.7/site-packages/ansible/runner/__init__.py", line 789, in _executor_internal
    return self._executor_internal_inner(host, self.module_name, self.module_args, inject, port, complex_args=complex_args)
  File "/usr/lib/python2.7/site-packages/ansible/runner/__init__.py", line 1013, in _executor_internal_inner
    complex_args = template.template(self.basedir, complex_args, inject, fail_on_undefined=self.error_on_undefined_vars)
  File "/usr/lib/python2.7/site-packages/ansible/utils/template.py", line 140, in template
    d[k] = template(basedir, v, templatevars, lookup_fatal, depth, expand_lists, convert_bare, fail_on_undefined, filter_fatal)
  File "/usr/lib/python2.7/site-packages/ansible/utils/template.py", line 124, in template
    varname = template_from_string(basedir, varname, templatevars, fail_on_undefined)
  File "/usr/lib/python2.7/site-packages/ansible/utils/template.py", line 382, in template_from_string
    res = jinja2.utils.concat(rf)
  File "<template>", line 9, in root
  File "/home/user/openshift-ansible/filter_plugins/openshift_master.py", line 476, in translate_idps
    IdentityProviderBase.validate_idp_list(idp_list)
  File "/home/user/openshift-ansible/filter_plugins/openshift_master.py", line 87, in validate_idp_list
    idp.validate()
  File "/home/user/openshift-ansible/filter_plugins/openshift_master.py", line 378, in validate
    if ('extraAuthorizeParameters' not in self.provider['extraAuthorizeParameters']
KeyError: 'extraAuthorizeParameters'
```

**Comment by detiber** [comment]
@anweiss I haven't had  a chance to test it yet (currently testing/troubleshooting a different issue), but I created the following PR to try to address this issue: https://github.com/openshift/openshift-ansible/pull/1592

**Comment by michaelgugino** [comment]
This issue appears to have been resolved by #1592
If this is still affecting you, please consider opening a new issue.
